File: backend/src/models/favorite.js
import { DataTypes } from "sequelize";
import db from "../config/database.js";
import User from "./user.js";
import Products from "./product.js";

const Favorite = db.define(
  "Favorite",
  {
    id: {
      type: DataTypes.INTEGER,
      primaryKey: true,
      autoIncrement: true,
    },
    userId: {
      type: DataTypes.INTEGER,
      allowNull: false,
      references: {
        model: "users", // Tetap gunakan string nama tabel
        key: "id",
      },
    },
    productId: {
      type: DataTypes.INTEGER,
      allowNull: false,
      references: {
        model: "products", // Tetap gunakan string nama tabel
        key: "id",
      },
    },
  },
  {
    timestamps: true,
    tableName: "favorites",
    createdAt: "created_at",
    updatedAt: "updated_at",
    indexes: [
      {
        unique: true,
        fields: ["userId", "productId"],
      },
    ],
  }
);

User.hasMany(Favorite, { foreignKey: "userId" });
Favorite.belongsTo(User, { foreignKey: "userId" });

Products.hasMany(Favorite, { foreignKey: "productId" });
Favorite.belongsTo(Products, { foreignKey: "productId" });

export default Favorite;

File: backend-go/models/favorite.go
package models

import "time"

type Favorite struct {
	ID        uint      `gorm:"primaryKey;autoIncrement"`
	UserID    uint      `gorm:"not null;index:idx_user_product,unique"`
	ProductID uint      `gorm:"not null;index:idx_user_product,unique"`
	CreatedAt time.Time `gorm:"autoCreateTime"`
	UpdatedAt time.Time `gorm:"autoUpdateTime"`

	// Relationships
	User    User     `gorm:"foreignKey:UserID;references:ID"`
	Product *Product `gorm:"foreignKey:ProductID;references:ID"`
}

func (Favorite) TableName() string {
	return "favorites"
}
